### PR TITLE
wrappers: update desktop helpers to support desktop IDs

### PR DESCRIPTION
**Pull request by ZeyadYasser**
`{Ensure,Remove}SnapDesktopFiles` are updated to support desktop file IDs.

Snaps are allowed to have their desktop files copied without name mangling (snap name prefix) if the file name is listed under `desktop-file-ids` attribute in their `desktop` interface declartion.

This is needed to support desktop IDs detailed in [SD170](https://docs.google.com/document/d/1j4flP303OJKrSxSCjkeZkh-9COuahVmo_EK_ye09RxQ/edit#heading=h.v7nluroo4fyr) spec.

### CODE REVIEW

**Comment by ZeyadYasser** [comment]
@alexmurray It would great to have a security review on this change, since files are being copied from the snap as snapd (root) without name mangling and we have been bitten by something similar before.

For context, we are not blindly copying desktop files without name mangling, they need to be approved by the store through a snap declaration https://github.com/canonical/snapd/pull/14277.

**Comment by ZeyadYasser** [comment]
@jhenstridge It would be great if you could take a look, I wasn't able to add you as a reviewer after the org migration, so pinging here instead.

File: wrappers/desktop_test.go
Comment (andrewphelpsj) [inline wrappers/desktop_test.go:133]: I think `expectedDesktopFilePath2` should just be `expectedDesktopFilePath` here? But this check is already superfluous, as the below `os.Stat` will error if it doesn't exist.

File: wrappers/desktop_test.go
Comment (andrewphelpsj) [inline wrappers/desktop_test.go:204]: It is better to use `const` for format strings. If you do that, then `go vet` will give you warnings when you make mistakes, rather than the string ending up weird at runtime.

```
phelps@arch ~/tmp> cat main.go
package main

import "fmt"

func main() {
        const template = "hello, %d\n"

        name := "andrew"
        fmt.Printf(template, name)
}
phelps@arch ~/tmp> go vet
# m
# [m]
./main.go:9:2: fmt.Printf format %d has arg name of wrong type string
```

File: wrappers/desktop_test.go
Comment (andrewphelpsj) [inline wrappers/desktop_test.go:343]: Double `Test` here.

**Review by andrewphelpsj — COMMENTED** [review]
Thanks you! Some minor comments, but looking good!

File: wrappers/desktop.go
Comment (MiguelPires) [inline wrappers/desktop.go:359]: I think this loop grew enough that `s` should be renamed to `info`

File: wrappers/desktop.go
Comment (MiguelPires) [inline wrappers/desktop.go:265]: This check and the type checks below aren't tested. May be worth a short table test

File: wrappers/desktop_test.go
Comment (ZeyadYasser) [reply wrappers/desktop_test.go:133]: Nice catch, will remove the check

File: wrappers/desktop_test.go
Comment (ZeyadYasser) [reply wrappers/desktop_test.go:204]: cool, I didn't know that

File: wrappers/desktop.go
Comment (ZeyadYasser) [reply wrappers/desktop.go:266]: Nice catch, added tests

**Review by MiguelPires — APPROVED** [review]
lgtm, thanks

**Review by andrewphelpsj — APPROVED** [review]
Thanks!

**Review by alexmurray — APPROVED** [review]
LGTM!

File: wrappers/desktop.go
Comment (pedronis) [inline wrappers/desktop.go:432]: and this can share code by having a  `forAllDesktopFiles(cb func(base, instaneName) error) ferror` unction that implements the looping?

File: wrappers/desktop.go
Comment (pedronis) [inline wrappers/desktop.go:386]: this and the same below could also be a helper

**Review by pedronis — COMMENTED** [review]
some small comments

File: wrappers/desktop.go
Comment (pedronis) [inline wrappers/desktop.go:405]: should this be called addGlobPatterAndConflictCheck ?

**Review by pedronis — APPROVED** [review]
thank you, small comment. There's been enough changes that you might want to get 1 re-review from somebody else

**Review by andrewphelpsj — APPROVED** [review]
Thanks!

**Pull request merged**
